Use QLatin1String instead of QStringLiteral when possible

rouault opened the pull request:
- Added  scripts/qstringfixup.sh and .py (manually run for now)
- QStringLiteral -> QLatin1String changes generated by scripts/qstringfixup.sh --all

nyalldawson commented on src/app/browser/qgsinbuiltdataitemproviders.cpp:
```diff
@@ -552,7 +552,7 @@ void QgsLayerItemGuiProvider::populateContextMenu( QgsDataItem *item, QMenu *men
     {
       // Also check for postgres layers (rasters are handled by GDAL)
       isFile = ( layerItem->providerKey() == QStringLiteral( "ogr" ) ||
```
Your script missed this line somehow

rouault replied:
fixed, as well as your other remarks

nyalldawson commented on src/app/browser/qgsinbuiltdataitemproviders.cpp:
```diff
@@ -552,7 +552,7 @@ void QgsLayerItemGuiProvider::populateContextMenu( QgsDataItem *item, QMenu *men
       isFile = ( layerItem->providerKey() == QStringLiteral( "ogr" ) ||
-                 layerItem->providerKey() == QStringLiteral( "gdal" ) ) &&
+                 layerItem->providerKey() == QLatin1String( "gdal" ) ) &&
                ! layerItem->uri().startsWith( QStringLiteral( "PG:" ) );
```
startsWith is another candidate you could catch too

nyalldawson commented on src/app/qgshandlebadlayers.cpp:
```diff
@@ -127,7 +127,7 @@ QgsHandleBadLayers::QgsHandleBadLayers( const QList<QDomNode> &layers )
     QString provider = node.namedItem( QStringLiteral( "provider" ) ).toElement().text();
     QString vectorProvider = type == QLatin1String( "vector" ) ? provider : tr( "none" );
-    bool providerFileBased = ( provider == QStringLiteral( "gdal" ) || provider == QStringLiteral( "ogr" ) || provider == QStringLiteral( "mdal" ) );
+    bool providerFileBased = ( provider == QStringLiteral( "gdal" ) || provider == QStringLiteral( "ogr" ) || provider == QLatin1String( "mdal" ) );
```
Here's some more missed opportunities

rouault commented:
@nyalldawson Since you reviewed it, I've also extended it to deal with != QStringLiteral() and .contains(QStringLiteral()) and .indexOf(QStringLiteral()). All have QLatin1String override, so I assume this is beneficial to replace them too

nyalldawson commented:
Was there any false positives here? Looks like a great candidate for including in the prepare commit script!

rouault commented:
> Was there any false positives here?

yes, there were unfortunately 2 instances of false positive (caught by adhoc `if m and 'qgetenv' not in line and 'h.first' not in line:`) for the == QStringLiteral("foo") or != QStringLiteral("foo") where the left-hand side was a QByteArray. If comparing a QByteArray with a QLatin1String, this results in an ambiguous situation for the compiler. For those very few situations, we could possibly require people to make it obvious by adding some QByteArray mention in the line, like ``/*QByteArray*/(x) == QStringLiteral("foo")`` and an enhanced script would skip those situations

But I'm realizing that https://github.com/KDE/clazy would probably a safer approach (I've never used it though), although it would require integration in the compilation.

rouault commented:
I've added logic to deal with QString::replace() as well. Run the script over tests/ as well
And I've added the script under scripts/ . Must be manually run for now

rouault committed:
Add scripts/qstringfixup.py and .sh

The .py script is the lower level one that acts on a single file.

The .sh one is the one aimed at being used by the user:
- qstringfixup.sh without argument fixes up files modified in git
- qstringfixup.sh --all runs on all .h/.cpp files under src/ and tests/

rouault commented:
Regarding QLatin1Char('x') vs 'x', yes I've hesitated. But if one day we would want to build with QT_NO_CAST_FROM_ASCII enabled, that would save fixing those places. But up to you, I can change to plain 'x'

rouault committed:
QStringLiteral -> QLatin1String changes generated by scripts/qstringfixup.sh --all
